### DISCUSSION TITLE
How to enable the address sanitizer on debug builds

### DISCUSSION THREAD

**Discussion by janvhs** [mesonbuild/meson#12760] [Q&A] [answered]
Hey, I want to enable the sanitizers in debug only. How would I acchive this? The method I tried doesn't work and I can't find any references in the manual or online.

This is what I am trying to do:

```meson
project(
  'Foo',
  'c',
  version: '0.0.1',
  license: 'GPL-3.0-or-later',
  meson_version: '>= 0.64.0',
  default_options: [
      'c_std=c2x',
      'warning_level=everything',
  ],
)

# This is the relevant part

if get_option('buildtype').startswith('debug')
  message('Building a debug release')
  add_project_arguments('b_sanitize=address,undefined', language: 'c')
else
  add_project_arguments(['werror=true'], language: 'c')
endif

ce_sources = file('src/main.c')

executable(
  meson.project_name().to_lower(),
  ce_sources,
)
```

Meson tries to add it as a link argument

**Comment by dcbaker** [comment, marked as answer]
Meson doesn't support setting global options inside the build definitions, since it mixes policy with build logic. The easiest way to handle this is to use a set of profile configurations:

debug.ini:
``ini
[built-in options]
buildtype = 'debug'
b_sanitize = 'address,undefined'
```
and release.ini:
```ini
[built-in options]
werror=true
```

then have a `profiles/` directory, you can run `meson setup builddir/debug --native-file profiles/debug.ini` and `meson setup builddir/release --native-file profiles/release.ini`

As an aside, I've had to professionally deal with third party code that mixes policy and logic like this, and as the downstream having to carry patches to pull that kind of policy out to get the project to build as the upstream continues to make changes to the code is incredibly frustrating.

**Reply by janvhs** [reply]
Thank you for the insight! Yeah today I wouldn't do something like this either but would set it via CLI arguments. The profile approach is arguably a way better option, so I'll adopt that. Thanks :D
